core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala

```diff
@@ -182,33 +197,55 @@ private[spark] class CoarseCookSchedulerBackend(
 
     val cmds = remoteConfFetch ++ environment ++ Seq(commandString, cleanup)
 
-    new Job.Builder()
+    val job = new Job.Builder()
       .setUUID(jobId)
       .setEnv(environment.asJava)
       .setUris(uris.asJava)
       .setCommand(cmds.mkString("; "))
-      .setMemory(sparkMesosScheduler.calculateTotalMemory(sc))
+      .setMemory(mesosBackend.calculateTotalMemory(sc))
       .setCpus(numCores)
       .setPriority(priority)
       .build()
+
+    executorsToJobs(slaveId.toString + "/" + taskId.toString) = jobId
+
+    job
   }
 
   def createRemainingJobs(): List[Job] = {
     def loop(coresRemaining: Int, jobs: List[Job]): List[Job] =
-      if (coresRemaining <= 0) jobs
-      else if (coresRemaining <= maxCoresPerJob) createJob(coresRemaining) :: jobs
-      else loop(coresRemaining - maxCoresPerJob, createJob(maxCoresPerJob) :: jobs)
-    loop(maxCores - totalCoresRequested, Nil).reverse
+      if (jobs.size == executorLimit) jobs
+      else
+        if (maxCoresPerJob <= 0) jobs
+        else if (coresRemaining <= maxCoresPerJob) createJob(coresRemaining) :: jobs
+        else loop(coresRemaining - maxCoresPerJob, createJob(maxCoresPerJob) :: jobs)
+
+    loop(maxCores - totalCoresRequested, List()).reverse
   }
 
   def requestRemainingCores() : Unit = {
     val jobs = createRemainingJobs()
-    totalCoresRequested += jobs.map(_.getCpus.toInt).sum
-    runningJobUUIDs = runningJobUUIDs ++ jobs.map(_.getUUID)
-    jobClient.submit(jobs.asJava, jobListener)
+
+    if (!jobs.isEmpty)
```

Comment:
I think this if is unnecessary as sum will return 0 if the list is empty, and ++ will concat a list with an empty list. What was your concern here?